### PR TITLE
Reliable Websocket

### PR DESCRIPTION
**Pull request by bpanthi977**
```
IMPORTANT: Please do not create a Pull Request without creating an issue first. 
Any change needs to be discussed and documented before proceeding.
Mention which issue(s) are closed by this PR below. 
For e.g "Fixes #123 Fixes #12" in case of multiple fixes or include the issues under "Linked Issues"
```

Fixes #

## Checklist 

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if necessary)
- [x] I have performed adequate tests that prove my fix is effective or that my feature works

## Proposed Changes
A reliable dropin replacement for websocket.
- Pings at a certain interval
- Retries lost connection
- Buffers messages during lost connection

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "reliable-websocket",
```
Comment (syaau) [inline package.json]: Can we have this named `@bhoos/websocket`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "version": "0.0.0",
+  "description": "reliable websocket client",
+  "main": "dist/index.js",
```
Comment (syaau) [inline package.json]: We want to go for es6 modules, so our typescript config should default to es6, and may be create a separate build for cjs.

```json
{
  "main": "dist/index.js",
  "module": "es6/index.js",
  "type": "module",
  "react-native": "src/index.js"
}
```

File: package.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  "author": "Bibek Panthi <bpanthi977@gmail.com>",
+  "license": "MIT",
+  "type": "module",
+  "scripts": {
+    "build": "rm -r dist/ && tsc",
+    "watch": "yarn build --watch"
+  },
+  "dependencies": {
+    "ring-buffer-ts": "^1.0.3",
+    "ws": "^8.8.1"
```
Comment (syaau) [inline package.json]: We will not depend on "ws" as it is not required on clients.

File: src/node.d.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { CloseEvent, Event, ErrorEvent, MessageEvent } from 'ws';
+import { ReliableWS as WS } from './index.js';
+
+export declare class ReliableWS extends WS {
+  onopen: ((event: Event) => void) | null;
+    onerror: ((event: ErrorEvent) => void) | null;
+    onclose: ((event: CloseEvent) => void) | null;
+    onmessage: ((event: MessageEvent) => void) | null;
+
+}
```
Comment (syaau) [inline src/node.d.ts]: @bpanthi977 Aren't we able to get rid of this file now ?

File: src/node.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import {WebSocket, Event, CloseEvent, MessageEvent, ErrorEvent} from 'ws';
+import { Config, ReliableWS as WS } from './index.js'
+//@ts-ignore
```
Comment (syaau) [inline src/node.ts:3]: Instead of ts-ignore we could define this on module level declaration

File: src/node.d.ts (hunk 4 above)
Comment (bpanthi977) [reply src/node.d.ts]: Yes we are. 
Oversight from my part.

File: src/node.ts (hunk 5 above)
Comment (bpanthi977) [reply src/node.ts:3]: I don't understand. Could you please explain.

I did it to please `tsc`. tsconfig has    
```json
    "lib": ["es2017", "DOM"],
```
so, tsc thinks that global.WebSocket is already assigned to browser's WebSocket. And thus reassigning it to ws WebSocket throws error.

Comment (syaau) [reply src/node.ts:3]: oh!, in that case, please ignore my comment on ts-ignore

**Review by syaau — APPROVED** [review]

**Pull request merged**